On migration from h2 to mysql, duplicate entry error with PermissionsGroup

Metabase version 0.20
on an aws ubuntu instance
error after running "java -jar metabase.jar load-from-h2"

`Transfering 3 instances of PermissionsGroup....BatchUpdateException:
 Message: Duplicate entry '3' for key 'PRIMARY'
 SQLState: 23000
 Error Code: 1062
java.sql.BatchUpdateException: Duplicate entry '3' for key 'PRIMARY'
	at com.mysql.jdbc.SQLError.createBatchUpdateException(SQLError.java:1166)
	at com.mysql.jdbc.PreparedStatement.executeBatchSerially(PreparedStatement.java:1773)
	at com.mysql.jdbc.PreparedStatement.executeBatchInternal(PreparedStatement.java:1257)
	at com.mysql.jdbc.StatementImpl.executeBatch(StatementImpl.java:958)
	at clojure.java.jdbc$execute_batch.invokeStatic(jdbc.clj:426)
	at clojure.java.jdbc$execute_batch.invoke(jdbc.clj:419)
	at clojure.java.jdbc$db_do_execute_prepared_statement$fn__2805.invoke(jdbc.clj:792)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:637)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:585)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:598)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:585)
	at clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:791)
	at clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:778)
	at clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:814)
	at clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:795)
	at clojure.java.jdbc$insert_cols_BANG_.invokeStatic(jdbc.clj:1069)
	at clojure.java.jdbc$insert_cols_BANG_.invoke(jdbc.clj:1061)
	at clojure.java.jdbc$insert_multi_BANG_.invokeStatic(jdbc.clj:1110)
	at clojure.java.jdbc$insert_multi_BANG_.invoke(jdbc.clj:1093)
	at metabase.cmd.load_from_h2$insert_entity_BANG_$fn__30770.invoke(load_from_h2.clj:125)
	at metabase.cmd.load_from_h2$insert_entity_BANG_.invokeStatic(load_from_h2.clj:124)
	at metabase.cmd.load_from_h2$insert_entity_BANG_.invoke(load_from_h2.clj:106)
	at metabase.cmd.load_from_h2$load_data_BANG_.invokeStatic(load_from_h2.clj:140)
	at metabase.cmd.load_from_h2$load_data_BANG_.invoke(load_from_h2.clj:134)
	at metabase.cmd.load_from_h2$load_from_h2_BANG_$fn__30850.invoke(load_from_h2.clj:209)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:613)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:585)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:629)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:585)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:598)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:585)
	at metabase.cmd.load_from_h2$load_from_h2_BANG_.invokeStatic(load_from_h2.clj:206)
	at metabase.cmd.load_from_h2$load_from_h2_BANG_.invoke(load_from_h2.clj:199)
	at clojure.lang.Var.invoke(Var.java:379)
	at metabase.core$fn__31589.invokeStatic(core.clj:209)
	at metabase.core$fn__31589.doInvoke(core.clj:207)
	at clojure.lang.RestFn.invoke(RestFn.java:397)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:646)
	at clojure.core$apply.invoke(core.clj:641)
	at metabase.core$run_cmd$fn__31592.invoke(core.clj:216)
	at metabase.core$run_cmd.invokeStatic(core.clj:216)
	at metabase.core$run_cmd.doInvoke(core.clj:211)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:648)
	at clojure.core$apply.invoke(core.clj:641)
	at metabase.core$_main.invokeStatic(core.clj:228)
	at metabase.core$_main.doInvoke(core.clj:224)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at metabase.core.main(Unknown Source)
Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Duplicate entry '3' for key 'PRIMARY'
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
	at com.mysql.jdbc.Util.getInstance(Util.java:387)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:934)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3966)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3902)
	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2526)
	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2673)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2549)
	at com.mysql.jdbc.PreparedStatement.executeInternal(PreparedStatement.java:1861)
	at com.mysql.jdbc.PreparedStatement.executeUpdateInternal(PreparedStatement.java:2073)
	at com.mysql.jdbc.PreparedStatement.executeBatchSerially(PreparedStatement.java:1751)
	... 51 more
Command failed with exception: Duplicate entry '3' for key 'PRIMARY'
`

--- a/src/metabase/core.clj
+++ b/src/metabase/core.clj
@@ -214,7 +214,8 @@
    (load-from-h2 nil))
   ([h2-connection-string]
    (require 'metabase.cmd.load-from-h2)
-   ((resolve 'metabase.cmd.load-from-h2/load-from-h2!) h2-connection-string)))
+   (binding [db/*disable-data-migrations* true]
+     ((resolve 'metabase.cmd.load-from-h2/load-from-h2!) h2-connection-string))))
 
 (defn ^:command profile
   "Start Metabase the usual way and exit. Useful for profiling Metabase launch time."

--- a/src/metabase/db.clj
+++ b/src/metabase/db.clj
@@ -339,6 +339,13 @@
     (format "Unable to connect to Metabase %s DB." (name engine)))
   (log/info "Verify Database Connection ... ✅"))
 
+
+(def ^:dynamic ^Boolean *disable-data-migrations*
+  "Should we skip running data migrations when setting up the DB? (Default is `false`).
+   There are certain places where we don't want to do this; for example, none of the migrations should be ran when Metabase is launched via `load-from-h2`.
+   That's because they will end up doing things like creating duplicate entries for the \"magic\" groups and permissions entries. "
+  false)
+
 (defn setup-db!
   "Do general preparation of database by validating that we can connect.
    Caller can specify if we should run any pending database migrations."
@@ -368,10 +375,10 @@
   ;; Establish our 'default' DB Connection
   (create-connection-pool! (jdbc-details db-details))
 
-  ;; Do any custom code-based migrations now that the db structure is up to date
-  ;; NOTE: we use dynamic resolution to prevent circular dependencies
-  (require 'metabase.db.migrations)
-  ((resolve 'metabase.db.migrations/run-all)))
+  ;; Do any custom code-based migrations now that the db structure is up to date.
+  (when-not *disable-data-migrations*
+    (require 'metabase.db.migrations)
+    ((resolve 'metabase.db.migrations/run-all))))
 
 (defn setup-db-if-needed!
   "Call `setup-db!` if DB is not already setup; otherwise this does nothing."

--- a/src/metabase/db/migrations.clj
+++ b/src/metabase/db/migrations.clj
@@ -218,6 +218,10 @@
                                (swap! processed-fields conj column-name)))))))))))))))))
 
 
+;;; +------------------------------------------------------------------------------------------------------------------------+
+;;; |                                                     PERMISSIONS v1                                                     |
+;;; +------------------------------------------------------------------------------------------------------------------------+
+
 ;; Add users to default permissions groups. This will cause the groups to be created if needed as well.
 (defmigration add-users-to-default-permissions-groups
   (let [{all-users-group-id :id} (perm-group/all-users)

